src/core/test/base/src/main/scala/TestBase.scala

```diff
@@ -204,3 +205,61 @@ abstract class TestBase extends FunSuite with BeforeAndAfterEachTestData with Be
   }
 
 }
+
+trait DataFrameEquality extends TestBase {
+  val epsilon = 1e-4
+  implicit lazy val doubleEq: Equality[Double] = TolerantNumerics.tolerantDoubleEquality(epsilon)
+  implicit lazy val dvEq: Equality[DenseVector] = new Equality[DenseVector]{
+    def areEqual(a: DenseVector, b: Any): Boolean = b match {
+      case bArr:DenseVector =>
+        a.values.zip(bArr.values).forall {case (x, y) => doubleEq.areEqual(x, y)}
+    }
+  }
+
+  implicit lazy val rowEq: Equality[Row] = new Equality[Row]{
+    def areEqual(a: Row, bAny: Any): Boolean = bAny match {
+      case b:Row =>
+        if (a.length != b.length) { return false }
+        (0 until a.length).forall(j =>{
+          a(j) match {
+            case lhs: DenseVector =>
+              lhs === b(j)
+            case lhs: Double if lhs.isNaN =>
+              b(j).asInstanceOf[Double].isNaN
+            case lhs: Double =>
+              b(j).asInstanceOf[Double] === lhs
+            case lhs =>
+              lhs === b(j)
+          }
+        })
+    }
+  }
+
+  val sortInDataframeEquality = false
+
+  val baseDfEq = new Equality[DataFrame]{
+    def areEqual(a: DataFrame, bAny: Any): Boolean = bAny match {
+      case ds:Dataset[_] =>
+        val b = ds.toDF()
+        if(a.columns !== b.columns){
```

Comment:
the spacing here is a little weird, eg there should be a space between if and (, ) and {.
Also same issues below - perhaps style cop wasn't run yet

Reply:
done